src/client/core/page-unload-barrier.js

```diff
@@ -0,0 +1,130 @@
+import hammerhead from './deps/hammerhead';
+import * as eventUtils from './utils/event';
+import { EventEmitter } from './utils/service';
+import delay from './utils/delay';
+
+var Promise      = hammerhead.Promise;
+var browserUtils = hammerhead.utils.browser;
+
+
+const DEFAULT_BARRIER_TIMEOUT       = 500;
+const WAIT_FOR_UNLOAD_TIMEOUT       = 3000;
+const SHORT_WAIT_FOR_UNLOAD_TIMEOUT = 30;
+
+const WAITING_FINISHED_EVENT = 'waiting-finished';
+
+
+class PageUnloadMonitor extends EventEmitter {
+    constructor () {
+        super();
+
+        this.waitingForUnload          = false;
+        this.waitingForUnloadTimeoutId = null;
+        this.unloading                 = false;
+
+        this._handleSubmit();
+        this._handleLinkClick();
+        this._handleBeforeUnload();
+    }
+
+    _overrideFormSubmit (form) {
+        var submit = form.submit;
+
+        form.submit = () => {
+            this._prolongUnloadWaiting(WAIT_FOR_UNLOAD_TIMEOUT);
+            submit.apply(form, arguments);
+        };
+    }
+
+    _handleSubmit () {
+        eventUtils.bind(document, 'submit', e => {
+            if (e.target.tagName.toLowerCase() === 'form')
+                this._prolongUnloadWaiting(WAIT_FOR_UNLOAD_TIMEOUT);
+        });
+
+        var forms = document.getElementsByTagName('form');
+
+        for (var i = 0; i < forms.length; i++)
+            this._overrideFormSubmit(forms[i]);
+    }
+
+    _handleLinkClick () {
```

Comment:
Code dublicate for https://github.com/DevExpress/testcafe-hammerhead/blob/master/src/client/sandbox/event/unload.js#L66